### PR TITLE
Adding check for cyclic configuration

### PR DESCRIPTION
**Pull request by HenningNT**
Not sure what happened to the previous PR, it got canceled automatically when I did the repository undo...
Anyway, here's a new fresh PR, which changes one thing only.
I added two tests to verify the the check throws ArgumentException, the reviewer might want to double check the test to verify that the created state machine is actually a cyclic configuration :-D

### CODE REVIEW

File: src/Stateless/StateConfiguration.cs [hunk 1]
```diff
@@ -508,6 +508,26 @@ public StateConfiguration OnExit(Action<Transition> exitAction, string exitActio
             /// <returns>The receiver.</returns>
             public StateConfiguration SubstateOf(TState superstate)
             {
+                var State = _representation.UnderlyingState;
+
+                // Check for accidental identical cyclic configuration 
+                if (State.Equals(superstate)) { throw new ArgumentException($"Configuring {State} as a substate of {superstate} creates an illegal cyclic configuration."); }
```
Comment (nblumhardt) [inline src/Stateless/StateConfiguration.cs]: Nit: apart from argument null checks, `if` statements in the repo are mostly over multiple lines. Braces optional, but if they're present they should be on a line each.

File: src/Stateless/StateConfiguration.cs [hunk 2]
```diff
@@ -508,6 +508,26 @@ public StateConfiguration OnExit(Action<Transition> exitAction, string exitActio
...
+
+                // Check for accidental identical nested cyclic configuration 
+                List<TState> superstates = new List<TState>();
```
Comment (nblumhardt) [inline src/Stateless/StateConfiguration.cs]: Nit: `var superstates = new List<TState>{ State };`

Comment (nblumhardt) [inline src/Stateless/StateConfiguration.cs]: Thinking through it a little further, could be:

```csharp
var superstates = new HashSet<TState>{ State };
```

The check for membership could then be moved inside the `while` loop as below.

File: src/Stateless/StateConfiguration.cs [hunk 3]
```diff
@@ -508,6 +508,26 @@ public StateConfiguration OnExit(Action<Transition> exitAction, string exitActio
...
+                superstates.Add(State);
+
+                // Build list of super states 
+                var activeRepresentation = _lookup(superstate);
+                while (activeRepresentation.Superstate != null)
+                {
+                    superstates.Add(activeRepresentation.Superstate.UnderlyingState);
```
Comment (nblumhardt) [inline src/Stateless/StateConfiguration.cs]: Using `HashSet` and moving the check here we can get away from the `GroupBy` below.

```csharp
if (superstates.Contains(activeRepresentation.Superstate.UnderlyingState))
    throw new ArgumentException($"Configuring {State} as a substate of {superstate} creates an illegal nested cyclic configuration.");
```

File: test/Stateless.Tests/StateMachineFixture.cs [hunk 4]
```diff
@@ -359,5 +359,20 @@ public void TheOnTransitionEventFiresBeforeTheOnEntryEvent()
                 Assert.AreEqual(expectedOrdering[i], actualOrdering[i]);
             }
         }
+        [Test]
+        public void CyclicConfigurationDetected()
```
Comment (nblumhardt) [inline test/Stateless.Tests/StateMachineFixture.cs]: `DirectCyclicConfigurationDetected()`?

File: test/Stateless.Tests/StateMachineFixture.cs [hunk 5]
```diff
@@ -359,5 +359,20 @@ public void TheOnTransitionEventFiresBeforeTheOnEntryEvent()
...
+        {
+            var sm = new StateMachine<State, Trigger>(State.A);
+
+            Assert.Throws(typeof(ArgumentException),  delegate { sm.Configure(State.A).SubstateOf(State.A); });
```
Comment (nblumhardt) [inline test/Stateless.Tests/StateMachineFixture.cs]: Is the old-style `delegate` needed?

```csharp
Assert.Throws(typeof(ArgumentException),  () => { sm.Configure(State.A).SubstateOf(State.A); });
```

File: test/Stateless.Tests/StateMachineFixture.cs [hunk 6]
```diff
@@ -359,5 +359,20 @@ public void TheOnTransitionEventFiresBeforeTheOnEntryEvent()
...
+        }
+        [Test]
```
Comment (nblumhardt) [inline test/Stateless.Tests/StateMachineFixture.cs]: Nit: missing newline.

**Comment by HenningNT** [comment]
Thanks for the feedback!
I've fixed the nitpicks, and added a necessary test (which failed initially).
Also switched to Hashset instead of list, it reduced the line count

**Comment by nblumhardt** [comment]
Nice! :-)

**Comment by HenningNT** [comment]
Thanks!

Sendt fra Windows 10-telefonen min

Fra: Nicholas Blumhardt
Sendt: søndag 22. januar 2017 kl. 23.46
Til: dotnet-state-machine/stateless
Kopi: HenningNT; Author
Emne: Re: [dotnet-state-machine/stateless] Adding check for cyclicconfiguration (#125)

Nice! :-)

**Pull request merged**